PR: Implement Profile Sidebar and New Setting Interface

bigtedde opened the pull request:
### Description
Vorta allows configuring profiles. A profile defines a number of sources, a backup destination, a schedule and some more settings. Currently there is a ComboBox for selecting a profile at the top of the window alongside buttons for adding or removing profiles. All profiles are active at the same time. However you can only edit the profile selected. This happens in the tabs below. The Misc tab is an exception. It contains the global settings that aren't associated to a profile.
In this task you will implement a sidebar next to the tab widget that replaces the current way of selecting a profile to edit. The sidebar should be comprised of a list with the configured profiles that can be selected and a visually distinct Misc item that gives access to the global settings. The old Misc tab will be removed in favour of this new user interface. The new Misc item now provides space for an about view and a way to manage profiles configured in Vorta.

Task outline: Implement the profile list that will be part of the sidebar. Add buttons for editing the profile list. Remove the old profile management GUI elements. Implement Misc item and the corresponding view. Move global settings to the new Misc view. Create an About tab and add it to the new Misc view. Implement a repository management tab and add it to the new Misc view. Write unittests for the new GUI parts.


### Related Issue
Addresses #1677 

### Motivation and Context
Easier management of profiles, separates the global settings tab from profile-specific tabs

### How Has This Been Tested?
Due to the nature of a main window redesign, some current tests will be broken until the final design is approved and I refactor the tests.

### Most Recent Demo:
![updated_sidebar](https://github.com/borgbase/vorta/assets/13988300/25622149-6be7-4b1e-a69c-6e9c634c3dcb)

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)




### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] I have read the [CONTRIBUTING](https://vorta.borgbase.com/contributing/) guide.
- [X] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.


*I provide my contribution under the terms of the [license](./../../LICENSE.txt) of this repository and I affirm the [Developer Certificate of Origin][dco].*

[dco]: https://developercertificate.org/

@m3nu @real-yfprojects 

<!--
This template is sourced from the awesome https://github.com/TalAter/open-source-templates
-->

bigtedde commented:
Current TODO:

- Refactor profile selector from ComboBox to a QListWidget (or something else?)
- Move Vorta/Borg versions + bug report link to "about" (maybe more items here, like link to Borg Docs?)
- Move profile export/rename to Profile Management tab (anything else should go here?)
- Create new tests for sidebar + settings panel, refactor old tests.

Anything else that should be added, or suggestions for improving the sidebar design/implementation please let me know :)

m3nu commented:
Screenshot looks good and there is space for a full list of profiles. 👍

real-yfprojects commented:
As I said I dought we need a profile management tab but a repe management tab could be useful.

Can you make it so that the settings button has different states when the settings are open or closed?

bigtedde commented:
> As I said I doubt we need a profile management tab but a repo management tab could be useful.

What features would we like to see in the repo management tab?

> Can you make it so that the settings button has different states when the settings are open or closed?

Sure - maybe have the button label change to " <- Home " or "Exit Settings"? Or simply fill it in with a background color?

bigtedde commented:
Here is the latest update. I created the "about" tab UI and changed the setting button when toggled. Thoughts?

![updated_sidebar](https://github.com/borgbase/vorta/assets/13988300/b69934c8-7744-424f-89ea-0991a16a1245)

m3nu commented:
I think you're on the right track. My thoughts:

- Set the profile list to expand all the way, so there is no vertical scrolling (same as similar Apple windows)
- Closing settings isn't needed IMO. When one clicks a profile, settings are gone anyways.
- Button *Settings* and then tab *Settings* is confusing. Maybe we want to have a few tabs for different subcategories of settings?
- What's "Repository Management" for? If this is a new feature, I'd keep it for later.

real-yfprojects commented:
> * Button _Settings_ and then tab _Settings_ is confusing.

Maybe `Misc` and `Settings`?



> * What's "Repository Management" for? If this is a new feature, I'd keep it for later.

It would be. The tab would resolve #1198.

bigtedde commented:
> Set the profile list to expand all the way, so there is no vertical scrolling (same as similar Apple windows)

I may need additional clarity here, or maybe an example of the Apple window you have in mind? I can allow the profile list to dynamically expand further down to accommodate more profiles, and this would cover most use cases, however with enough profiles there would inevitably be some vertical scrolling.

> Closing settings isn't needed IMO. When one clicks a profile, settings are gone anyways.

This is not the current behavior, but I could implement it. I'm open to other ideas too, maybe a 'back' or 'home' button in the top left of the interface that always takes users back to the repository tab?

> Button Settings and then tab Settings is confusing. Maybe we want to have a few tabs for different subcategories of settings?

@real-yfprojects idea of a `MIsc` button and then a `Setting` panel could solve this overlap.

real-yfprojects commented:
> This is not the current behavior, but I could implement it.

That would be great!

> I may need additional clarity here

You can make the list expand automatically filling the room that currently resides between the list and the settings button.

bigtedde commented:
> > This is not the current behavior, but I could implement it.
> 
> That would be great!
> 
> > I may need additional clarity here
> 
> You can make the list expand automatically filling the room that currently resides between the list and the settings button.

I implemented both of those features this morning. Check it out and LMK when you think!

Also, regarding the repo management tab - did we want this to be a separate PR, or combine it here?

real-yfprojects commented:
Can you swap the trash can through a minus sign? And right align the plus and minus sign while left aligning the others?

real-yfprojects commented:
> Also, regarding the repo management tab - did we want this to be a separate PR, or combine it here?

I would rather leave that for a separate PR.


Regarding the about view can you add the following?

* Copyright
* License
* Link Repo and Docs

bigtedde commented:
> Can you swap the trash can through a minus sign? And right align the plus and minus sign while left aligning the others?

Done 👍 

> Regarding the about view can you add the following?
> 
> * Copyright
> * License
> * Link Repo and Docs

Certainly. 

- Where should I link to for copyright? Could not find copyright info on GitHub or vorta.borgbase.com
- For license, should I link to GitHub or open the license.txt file?

real-yfprojects commented:
```
  <one line to give the program's name and a brief idea of what it does.>

Copyright (C) 2018-2020 Manuel Riel and Vorta contributors (see <CONTRIBUTORS.md>)

     Vorta is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, version 3 only.

    This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

    You should have received a copy of the GNU General Public License along with this program. If not, see <https://www.gnu.org/licenses/>. 
```

Then you could also add one of [the GPL logos](https://www.gnu.org/graphics/license-logos.html) and a [python logo](https://www.python.org/community/logos/).

bigtedde commented:
After talking with @m3nu, a few ideas came up for discussion:

1) keeping the profile selector a set height (400px) instead of increasing/decreasing in size based on number of profiles.
2) renaming the "misc" button to a more explicit "Settings / About" button
3) removing the "repository management" tab from this PR, out of scope
4) Adding the contributions, license, and repo/docs links info to the about tab
5) Have the style of profile action buttons mimic the style found here:
![image](https://github.com/borgbase/vorta/assets/13988300/07f01a9b-eaed-4f51-b262-cd35afa9b6b6)

@real-yfprojects Do you have any feedback/ideas for these topics?

real-yfprojects commented:
> 1. keeping the profile selector a set height (400px) instead of increasing/decreasing in size based on number of profiles.

Actually, I would make the size dependent on the window size. Basically like it is done in the screenshot of the network dialog you shared.




> 4\. Adding the contributions, license, and repo/docs links info to the about tab

Didn't we already decide to do that?



> 5\. Have the style of profile action buttons mimic the style found here:

Not sure how this will look in Qt, but feel free to share a screenshot once possible.

bigtedde commented:
Latest look:

![new_sidebar_gif](https://github.com/borgbase/vorta/assets/13988300/8484f3a5-1b26-4ea5-ba7f-1231cb6b3a59)

With started a new job these last two weeks, I have found myself in a position with less time than originally anticipated, and so I am considering cutting the scope of just a couple things:

- Making the profile selector resize with the window. I tried a couple approaches, but led to unintended behavior in other areas so I have left the profile selector a fixed size for now. 

- Making the profile action buttons in-line with the profile selector. Once again I tried a couple approaches but Qt doesn't have any built in ways to put the buttons inside a widget.

bigtedde commented:
I have updated the current tests to work with the new profile sidebar, and next I will add in a few tests for the "about" tab as well as the settings button to ensure it is always opening/closing the settings panel correctly.

m3nu commented:
Looks great! Trying this out right now.

m3nu commented:
Works well and it's a nice refresh. Well done!

I'd maybe play with the button order a bit to go from most to least frequent. And Plus on the left is also more common. Maybe like this: `+ -   Rename Export`?

And possible to allow changing the size between profile list and the rest of the app? One may want more space on either side.

And for leaving the *Settings* view: Currently you click either the same button again or change the profile? If there is only one profile (like when first using the app), the latter option isn't possible. So this could make sense:

1. Subtly changing the *Settings/About* button state to show it's active (Qt may have a setting for this already)? So it's clear it can be clicked again. I know we changed this from having an explicit *Close* label before. Maybe something in the middle will work well.
2. Going back to *Profile* view when clicking the profile list, without changing the profile?

Not sure those make sense. Would love for others to weigh in too. Goal is basically to make closing the *Settings* view more intuitive.

bigtedde commented:
> I'd maybe play with the button order a bit to go from most to least frequent. And Plus on the left is also more common. Maybe like this: + -   Rename Export?

Cool, I can move these around 👍 

> And possible to allow changing the size between profile list and the rest of the app? One may want more space on either side.

I'll take a look into this. Would want to ensure I could implement without much impact on the layout of other  elements of the app.

> Subtly changing the Settings/About button state to show it's active (Qt may have a setting for this already)? So it's clear it can be clicked again. I know we changed this from having an explicit Close label before. Maybe something in the middle will work well.

How would you feel about filling in the background of the button blue to indicate it is selected, similar to how we fill in the background of the tab widget to indicate the selected tab?

> Going back to Profile view when clicking the profile list, without changing the profile?

Yes, good idea and I was thinking about this ideas as well. Should be straight forward enough to implement, and would be good for large amount of users with a single profile.

bigtedde commented:
@m3nu Added changes:

- rearranged profile action buttons.
- added a blue background to button when user is in the "settings / about" tab.
- selecting the current profile will now return to the main screen if the settings / about tab is open.

Next I will work to allow changing the size between profile list and the rest of the app, and then I will write tests for the new features.

m3nu commented:
Hey Ted, thanks for those latest changes. I think it's coming along nicely and gives it a new and polished look.

Regarding the button state, we need to be careful with colors due to different skins in Qt, as well as dark/light node. Maybe just make the button font bold, if that's possible? I couldn't find a suitable "button active" state in the docs.

bigtedde commented:
> Maybe just make the button font bold, if that's possible?

I implemented this change 👍

bigtedde commented:
Marked as ready for review. Slightly reduced scope from the original outline, but this profile sidebar should make our users happy. 

After this is merged into master, I will create a follow-up post about features that remain to be implemented.

m3nu commented:
Work well here. Can be merged. What do you think @real-yfprojects @Hofer-Julian ? Some screenshots:

![Screenshot 2023-10-20 at 10 38 03](https://github.com/borgbase/vorta/assets/3916435/218fc3ee-68a8-40bc-a30b-39e7e58b7a39)
![Screenshot 2023-10-20 at 10 38 09](https://github.com/borgbase/vorta/assets/3916435/1ed34a81-25cb-4d60-ae32-d13bd0d87db0)
![Screenshot 2023-10-20 at 10 38 15](https://github.com/borgbase/vorta/assets/3916435/d56978af-8d3f-43ef-b72b-b793df3d28a2)

m3nu commented:
Is `close_X.svg` still used? If not it can be removed from the PR.

Where is `settings_wheel.svg` from? It has some comments with "SVG mixer" ? If it's from a random site, it would be better to use a similar icon from ForkAwesome or Google Material. Since we just cleaned up our icon licenses. More here: https://vorta.borgbase.com/contributing/#icons

real-yfprojects commented:
Looks pretty cool! That's exactly what I had in mind when added this task to the ideas page.

real-yfprojects commented on src/vorta/views/utils.py:
```diff
@@ -15,3 +15,17 @@ def get_colored_icon(icon_name):
     return QIcon(QPixmap(svg_img))
+
+
+def get_colored_pixmap(icon_name, scaled_height):
+    """
+    Return SVG icon in the correct color.
+    """
+    with open(get_asset(f"icons/{icon_name}.svg"), 'rb') as svg_file:
+        svg_str = svg_file.read()
+    if uses_dark_mode():
+        svg_str = svg_str.replace(b'#000000', b'#ffffff')
+    # Reduce image size to specified scaled height
+    svg_img = QImage.fromData(svg_str).scaledToHeight(scaled_height)
+
+    return QPixmap(svg_img)
```
It would make sense to alter `get_colored_icon` so that it uses this new method.

bigtedde replied:
I added a couple optional args in `get_colored_icon ` to accomplish this 👍

Hofer-Julian reviewed (approved):
Screenshots look good to me

bigtedde commented:
@m3nu removed extra icon, switched to Material for the settings_wheel, and removed the extra utils function @real-yfprojects pointed out. Please let me know if there is anything else to button up before merge 👍

real-yfprojects commented on src/vorta/views/utils.py:
```diff
@@ -3,7 +3,7 @@
 
 
-def get_colored_icon(icon_name):
+def get_colored_icon(icon_name, scaled_height=120, return_qpixmap=False):
```
You now scale to 120, where previously the default was 128. Has that a reason?

bigtedde replied:
Sorry - typo. Fix pushed :)

m3nu commented:
Nothing left to add from me. If nobody objects, I'll merge this and do a new beta-release to get even broader testing.

Thanks all!

m3nu commented:
This concludes @bigtedde's GSoC projects. Thanks again for the great work and for extending to finish everything.
